x/pool-incentives/keeper/distr_test.go

```diff
@@ -300,6 +302,132 @@ func (s *KeeperTestSuite) TestReplaceDistrRecords() {
 	}
 }
 
+func (s *KeeperTestSuite) TestReplaceDistrRecords_GroupGauge() {
+	s.SetupTest()
+	keeper := s.App.PoolIncentivesKeeper
+
+	poolGroup := s.PrepareAllSupportedPools()
+
+	// Create perpetual group
+	perpetualGroupGauge, err := s.App.IncentivesKeeper.CreateGroupAsIncentivesModuleAcc(s.Ctx, 0, []uint64{poolGroup.BalancerPoolID, poolGroup.ConcentratedPoolID, poolGroup.StableSwapPoolID})
+	s.Require().NoError(err)
+
+	// Create non-perpetual group
+	nonPerpetualGroupGauge, err := s.App.IncentivesKeeper.CreateGroupAsIncentivesModuleAcc(s.Ctx, 1, []uint64{poolGroup.BalancerPoolID, poolGroup.ConcentratedPoolID, poolGroup.StableSwapPoolID})
+	s.Require().NoError(err)
+
+	// Initial state to use to ensure replace actually removes previous records.
+	initialDistrRecords := []types.DistrRecord{
+		{
+			GaugeId: 0,
+			Weight:  osmomath.NewInt(500),
+		},
+		{
+			GaugeId: 1,
+			Weight:  osmomath.NewInt(600),
+		},
+		{
+			GaugeId: 2,
+			Weight:  osmomath.NewInt(700),
+		},
+	}
+
+	tests := []struct {
+		name               string
+		testingDistrRecord []types.DistrRecord
+		expectedErr        error
+		expectTotalWeight  osmomath.Int
+	}{
+		{
+			name: "Perpetual group",
+			testingDistrRecord: []types.DistrRecord{
+				{
+					GaugeId: 0,
+					Weight:  osmomath.NewInt(100),
+				},
+				{
+					GaugeId: perpetualGroupGauge,
+					Weight:  osmomath.NewInt(100),
+				},
+			},
+			expectTotalWeight: osmomath.NewInt(200),
+		},
+		{
+			name: "Perpetual group, new weights",
+			testingDistrRecord: []types.DistrRecord{
+				{
+					GaugeId: 0,
+					Weight:  osmomath.NewInt(100),
+				},
+				{
+					GaugeId: perpetualGroupGauge,
+					Weight:  osmomath.NewInt(200),
+				},
+			},
+			expectTotalWeight: osmomath.NewInt(300),
+		},
+		{
+			name: "Error: non perpetual group",
+			testingDistrRecord: []types.DistrRecord{
+				{
+					GaugeId: 0,
+					Weight:  osmomath.NewInt(100),
+				},
+				{
+					GaugeId: nonPerpetualGroupGauge,
+					Weight:  osmomath.NewInt(200),
+				},
+			},
+			expectedErr: errorsmod.Wrapf(types.ErrDistrRecordRegisteredGauge,
+				"Gauge ID #%d is not perpetual.",
+				nonPerpetualGroupGauge),
+		},
+		{
+			name: "Error: perpetual and non perpetual group",
+			testingDistrRecord: []types.DistrRecord{
+				{
+					GaugeId: 0,
+					Weight:  osmomath.NewInt(100),
+				},
+				{
+					GaugeId: perpetualGroupGauge,
+					Weight:  osmomath.NewInt(100),
+				},
+				{
+					GaugeId: nonPerpetualGroupGauge,
+					Weight:  osmomath.NewInt(100),
+				},
+			},
+			expectedErr: errorsmod.Wrapf(types.ErrDistrRecordRegisteredGauge,
+				"Gauge ID #%d is not perpetual.",
+				nonPerpetualGroupGauge),
+		},
+	}
+
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			// Set up distribution records prior to replace, to ensure replace deletes them.
+			err := keeper.ReplaceDistrRecords(s.Ctx, initialDistrRecords...)
```

Comment:
Is there another way we could set up this test case?

It might get confusing if bug occurs and we have system under test setting up system under test

Reply:
Yeah good call, was thinking about this but your comment convinced me. Let me think about another way to set it up

Reply:
https://github.com/osmosis-labs/osmosis/pull/6619/commits/204bc57ece670331179c2b5c71ace1092725f7ea